### PR TITLE
Unsend Requests

### PR DESCRIPTION
**Pull request by RyanRory**
- Use the `delete` endpoint to delete a specific message with the hash
- Send `UnsendRequest` to the recipient and linked devices to delete the message
- Dismiss the notification of a deleted message
- New action sheet UI for delete options
- Only show the deletion option when all selected messages are outgoing
- New message bubble UI for deleted messages

### CODE REVIEW

File: app/src/main/res/layout/fragment_delete_message_bottom_sheet.xml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    android:layout_height="wrap_content"
+    android:orientation="vertical"
+    app:behavior_hideable="true"
+    app:layout_behavior="com.google.android.material.bottomsheet.BottomSheetBehavior">
+
+    <TextView
+        android:id="@+id/deleteForMeTextView"
+        style="@style/BottomSheetActionItem"
+        android:text="@string/delete_message_for_me"
+        android:textColor="@color/core_red"/>
```
Comment (nielsandriesse) [inline app/src/main/res/layout/fragment_delete_message_bottom_sheet.xml:16]: We have our own brand red, which is called `@color/destructive` :)

File: app/src/main/java/org/thoughtcrime/securesms/conversation/v2/ConversationActivityV2.kt [hunk 2]
```diff
@@ -1114,7 +1121,59 @@ class ConversationActivityV2 : PassphraseRequiredActionBarActivity(), InputBarDe
         stopAudioHandler.removeCallbacks(stopVoiceMessageRecordingTask)
     }
 
-    override fun deleteMessages(messages: Set<MessageRecord>) {
+    private fun buildUsendRequest(message: MessageRecord): UnsendRequest? {
```
Comment (nielsandriesse) [inline app/src/main/java/org/thoughtcrime/securesms/conversation/v2/ConversationActivityV2.kt:1124]: `buildUnsendRequest`

**Comment by hjubb** [comment]
I think this needs to be changed to merge against dev

**Review by hjubb — CHANGES_REQUESTED** [review]
looks good, just those couple comments and changing the base to `oxen-io:dev` instead of master

File: app/src/main/java/org/thoughtcrime/securesms/attachments/DatabaseAttachmentProvider.kt [hunk 3]
```diff
@@ -167,14 +168,28 @@ class DatabaseAttachmentProvider(context: Context, helper: SQLCipherOpenHelper)
...
+                                                   else DatabaseFactory.getMmsDatabase(context)
+        messagingDatabase.deleteMessage(messageID)
         DatabaseFactory.getLokiMessageDatabase(context).deleteMessage(messageID, isSms)
+        DatabaseFactory.getLokiMessageDatabase(context).deleteMessageServerHash(messageID)
+    }
+
+    override fun updateMessageAsDeleted(timestamp: Long, author: String) {
+        val database = DatabaseFactory.getMmsSmsDatabase(context)
+        val address = Address.fromSerialized(author)
+        val message = database.getMessageFor(timestamp, address)!!
```
[the remark this answers is not in the document]
Comment (RyanRory) [reply app/src/main/java/org/thoughtcrime/securesms/attachments/DatabaseAttachmentProvider.kt:181]: Yeah, we can return here. Theoretically, this function shouldn't be called if the message id is null. The message id is checked before this function is called. But we should return here to be safe.

**Review by hjubb — APPROVED** [review]
code looks good, would be fine to merge if there's no outstanding bugs / issues :+1:

**Pull request merged**